frontend/src/features/admin-form/responses/ResponsesPage/storage/useDecryptionWorkers.ts

```diff
@@ -189,43 +183,57 @@ const useDecryptionWorkers = ({
                   case CsvRecordStatus.Ok: {
                     try {
                       csvGenerator.addRecord(decryptResult.submissionData)
-                      receivedRecordCount++
                     } catch (e) {
                       errorCount++
                       console.error('Error in getResponseInstance', e)
                     }
+                  }
+                }
+                return decryptResult
+              })
+              .then((decryptResult) => {
+                decryptionProgress += 1
+                onProgress(decryptionProgress)
+                return decryptResult
+              }),
+          )
+          currentSubmissionIndex += 1 // used to assign the next submission to the next worker
+          return reader.read().then(read)
+        }),
+      )
 
-                    if (downloadAttachments && decryptResult.downloadBlob) {
-                      // Ensure attachments downloads are spaced out to avoid browser blocking downloads
-                      if (progress % ATTACHMENT_DOWNLOAD_CONVOY_SIZE === 0) {
-                        const now = new Date().getTime()
-                        const elapsedSinceXDownloads =
-                          now - timeSinceLastXAttachmentDownload
-
-                        const waitTime = Math.max(
-                          0,
-                          ATTACHMENT_DOWNLOAD_CONVOY_MINIMUM_SEPARATION_TIME -
-                            elapsedSinceXDownloads,
-                        )
-                        if (waitTime > 0) {
-                          await waitForMs(waitTime)
-                        }
-                        timeSinceLastXAttachmentDownload = now
-                      }
-                      await downloadResponseAttachment(
-                        decryptResult.downloadBlob,
-                        decryptResult.id,
-                      )
-                    }
+      return new Promise<DownloadResult>((resolve, reject) => {
+        // Step 1: Decrypt all submissions and their attachments (into blobs), add the submissions into the csv.
+        Promise.all(submissionDecryptPromises)
```

Comment:
question: should this be `Promise.allSettled` instead?
- it looks like we'd want to handle decryption of any single submission gracefully, rather than halt completely.
- (I recognise that this is part of a larger changeset of PRs that may address this elsewhere)

Reply:
great catch. i have updated such that the file saving should be done after the submission is decrypted to replicate previous behavior. 

this prevents the case where: attachment files are not yet saved to device but the download modal's `onProgress` is already updated to 100% and hence the download progress is out of sync with actual file downloading. 

with the new implementation, `Promise.all` is used to reject if any one of the submissions fail, so that the catch will now do the necessary logs and killing of workers